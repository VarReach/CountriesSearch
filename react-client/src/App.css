html {
  box-sizing: border-box;
}

body {
  margin: 0;
}

#root {
  background-color: #221f1f;
}

* {
  box-sizing: inherit;
  font-family: 'Open Sans', sans-serif;
  color: #f9f9f9;
}

.hide {
  display: none !important;
}