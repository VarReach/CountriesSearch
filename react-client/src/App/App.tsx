import React, { FC } from "react";

export const App: FC = () => {
  return (<div>Testing</div>);
}

export default App;