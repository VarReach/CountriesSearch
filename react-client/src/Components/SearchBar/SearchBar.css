.search_form {
  position: relative;
  background-color: #3e3a3a;
  padding: 0.1em 12.6em 0.1em 0.8em;
  border-radius: 3px;
  /* overflow: hidden; */
  width: 28em;
  border-color: #141414;
  border-style: solid;
  border-width: 0 1px 2px 1px;

  transition: width 0.2s ease-in-out;
}
.search_form.code {
  width: 18em;
}
.search_form.code .search_form__input {
  letter-spacing: 0.2em;
  text-transform: uppercase;
}

.search_form.error .search_form__input {
  border-bottom: 1px solid #e45454;
}

.search_form__input {
  font-size: 1.05em;
  font-weight: normal;
  background: none;
  outline: none;
  border: none;
  padding: 0;
  height: 2.4em;
  width: 100%
}

.search_form__inputs {
  position: absolute;
  right: 0;
  top: 0;
  bottom: 0;
  display: flex;
  z-index: 1;
}

.search_form__button {
  display: inline-flex;
  justify-content: center;
  align-items: center;
  background: none;
  border: none;
  outline: none;
  cursor: pointer;
}
.search_form__button * {
  color: #9f9b9b;
}
.search_form__button:hover *,
.search_form__button:focus * {
  color: #c5c1c1;
}

.search_form__type-button,
.search_form__submit {
  border-left: 1px solid #524e4e;
  font-size: 1.05em;
  height: 2.6em;
}

.search_form__type-button:hover,
.search_form__submit:hover,
.search_form__type-button:focus,
.search_form__submit:focus {
  background-color: #524e4e;
}

.search_form__submit {
  width: 2.6em;
  border-top-right-radius: 3px;
  border-bottom-right-radius: 3px;
}
.search_form__submit.highlight {
  background-color: #64a23a;
}
.search_form__submit.highlight * {
  color: #d6d3d3;
}
.search_form__submit.highlight:hover * {
  color: #f9f9f9;
}

.search_form__type-button {
  padding-left: 0.6em;
  width: 7.3em;
  justify-content: space-between;
}
.search_form__type-button span {
  font-size: 1em;
  text-transform: capitalize;
}

.search_form__clear {
  position: absolute;
  right: 12.6em;
  top: 0;
  bottom: 0;
}
.search_form__clear i {
  font-size: 1.2em !important;
}

.search_form__error {
  position: absolute;
  top: 100%;
  left: 2.2em;
  font-size: 0.8em;
  color: #e45454;
}


/** Type Dropdown */

.search-type-dropdown {
  position: absolute;
  top: 100%;
  right: 2.4em;
  width: 8em;
  margin-top: 0.2em;
  background-color: #726d6d;
  border-color: #141414;
  border-style: solid;
  border-width: 0 1px 3px 1px;
  border-radius: 5px;
  overflow: hidden;
}

.search-type-dropdown__list {
  list-style: none;
  padding: 0;
  margin: 0;
}

.search-type-dropdown__list-item,
.search-type-dropdown__button {
  width: 100%;
}

.search-type-dropdown__button {
  outline: none;
  border: none;
  background: none;
  cursor: pointer;
  text-align: left;
  text-transform: capitalize;
  height: 2em;
}


.search-type-dropdown__button:not(.selected):hover {
  background-color: #524e4e;
}
.search-type-dropdown__button.selected {
  background-color: #64a23a;
}