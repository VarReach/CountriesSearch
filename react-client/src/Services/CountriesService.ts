import { Url } from "url";
import axios from "axios";

const BASE_URL = `https://${window.location.hostname}:${window.location.port}/api/countries`;

export type SearchType = "code" | "name" | "fullName"

export interface Country {
  languages: {
    iso639_1: string,
    iso639_2: string,
    name: string,
    nativeName: string,
  }[],
  flag: Url,
  name: string,
  alpha2code: string,
  alpha3code: string,
  region: "AFRICA" | "AMERICAS" | "ASIA" | "EUROPE" | "OCEANIA",
  subregion: string,
  population: number,
};

function CountriesService() {
  const service = {
    get,
  };
  // ==================
  //   PUBLIC METHODS
  // ==================
  function get(type: SearchType, value: string): Promise<Country[]> {
    const queryString = `?type=${type}&value=${value}`;
    return axios.get(BASE_URL + queryString);
  }

  return service;
}

export default CountriesService();