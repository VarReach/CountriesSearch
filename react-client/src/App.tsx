import React, { FC, useState, useEffect } from "react";
import CountriesService, { Country, SearchType } from "./Services/CountriesService";
import QS from "query-string";
// Components
import MainPage from "./Components/MainPage";
import SearchPage from "./Components/SearchPage";

type SearchContextValue = {
  fetchCountries: (type: SearchType, value: string) => void;
}
const SearchContext = React.createContext<SearchContextValue>({ fetchCountries: () => null });

export const App: FC = () => {
  const [countries, setCountries] = useState<Country[]>();
  const [error, setError] = useState<Error>();
  const [loading, setLoading] = useState<boolean>(false);

  const fetchCountries = async (type: SearchType, value: string) => {
    if (!loading) setLoading(true);
    try {
      const countries = await CountriesService.get(type, value);
      setCountries(countries);
    } catch (err) {
      setError(err);
    } finally {
      setLoading(false);
    }
  }

  const updateURL = (queries: { type?: SearchType, value?: string}) => {
    const queryString = QS.stringify(queries);
    const newUrl = new URL(window.location.href);
    newUrl.search = `?${queryString}`;
    window.history.pushState(null, "", newUrl.href);
  }

  useEffect(() => {
    // Check for valid Querie params -> execute the search
    if (window.location.search) {
      // Ignore unrelated query params
      const queries = QS.parse(window.location.search);
      const { type } = queries;
      let { value } = queries;
      // Check to make sure Type is valid, if it's not, clear it
      if (type !== "code" && type !== "name" && type !== "fullName") {
        updateURL({ ...queries, type: undefined });
      } else if (type && value) {
        // Use last value if multiple are provided
        value = Array.isArray(value) ? value[value.length - 1] : value;
        // Remove additional values from URL
        updateURL({ ...queries, value });
        fetchCountries(type, value);
      }
    }
  }, []);

  return (
    <SearchContext.Provider value={{ fetchCountries }}>
      {loading || (countries && countries.length) ? 
      <SearchPage /> : 
      <MainPage />
      }
    </SearchContext.Provider>
  );
}

export default App;