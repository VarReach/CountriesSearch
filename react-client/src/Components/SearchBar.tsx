import React, { FC, useState } from "react";
import { SearchType } from "../Services/CountriesService";

interface SearchBarProps {
  page: "main" | "results";
}

export const SearchBar: FC<SearchBarProps> = (props) => {
  const { page } = props;
  const [searchValue, setSearchValue] = useState<string>("");
  // Default to 'name' search type
  const [searchType, setSearchType] = useState<SearchType>("name");

  const updateSearchValue = (event: React.ChangeEvent<HTMLInputElement>) => {
    event.preventDefault();
    setSearchValue(event.target.value);
  }

  return (
    <form id={`search_form__${page}`} className={`search_${page}`} onSubmit={(event) => event.preventDefault()}>
      <input 
        id={`search_form_input__${page}`} 
        className={`search_input__${page}`} 
        tabIndex={1} 
        value={searchValue} 
        autoComplete="off" 
        autoCapitalize="off" 
        autoCorrect="off" 
        onChange={updateSearchValue}
      />
      <input
        id={`search_form_button__${page}`}
        className={`search_button__${page}`}
        type="submit"
        tabIndex={2}
      />
    </form>
  );
}

export default SearchBar;