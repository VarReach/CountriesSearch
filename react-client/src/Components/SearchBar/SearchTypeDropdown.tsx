import React, { FC, useEffect, useCallback } from "react";
import { SearchType } from "../../Services/CountriesService";

interface SearchTypeDropdownProps {
  page: "main" | "results";
  visible: boolean;
  selectedType: SearchType;
  switchType: (type: SearchType) => void;
  setVisible: (visible: boolean) => void;
}


export const SearchTypeDropdown: FC<SearchTypeDropdownProps> = (props) => {
  const { page, selectedType, visible, setVisible, switchType } = props;

  const options = ["name", "fullName", "code"]

  const clickEvent = useCallback(() => {
    setVisible(false);
  }, []);
  const keyDownEvent = useCallback((event) => {
    if (event.key === "Escape") setVisible(false);
  }, []);

  /**
   * Hide the dropdown when the user clicks outside of it or presses the 'escape' key
   */
  useEffect(() => {
    if (visible) {
      document.addEventListener("click", clickEvent);
      document.addEventListener("keydown", keyDownEvent);
    } else {
      document.removeEventListener("click", clickEvent);
      document.removeEventListener("keydown", keyDownEvent);
    }
  }, [visible])

  // Make sure event listeners are removed on dismount
  useEffect(() => {
    return () => {
      document.removeEventListener("click", clickEvent);
      document.removeEventListener("keydown", keyDownEvent);
    }
  }, []);

  if (visible) {
    return (
      <div id={`search-type-dropdown__${page}`} className="search-type-dropdown" role="listbox">
        <ul className="search-type-dropdown__list">
          {options.map((option, i) => {
            return (
              <li key={i} className="search-type-dropdown__list-item">
                <button 
                  type="button"
                  title={option}
                  className={
                    "search-type-dropdown__button" + (selectedType === option ? " selected" : "")
                  }
                  onClick={() => switchType(option as SearchType)}
                >
                  {option === "fullName" ? "full name" : option}
                </button>
              </li>
            );
          })}
        </ul>
      </div>
    );
  } else {
    return null;
  }
}

export default SearchTypeDropdown;