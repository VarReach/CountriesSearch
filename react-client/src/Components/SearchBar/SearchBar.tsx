import React, { FC, useState, useContext, useRef } from "react";
import { SearchType } from "../../Services/CountriesService";
import "./SearchBar.css";
// Context
import { SearchContext } from "../../App";
// Components
import SearchTypeDropdown from "./SearchTypeDropdown";

export interface SearchBarProps {
  page: "main" | "results";
}

export const SearchBar: FC<SearchBarProps> = (props) => {
  const { page } = props;
  const {
   fetchCountries, 
   searchValue, 
   setSearchValue, 
   searchType, 
   setSearchType 
  } = useContext(SearchContext);
  // Default to 'name' search type

  const [submitError, setSubmitError] = useState<boolean>(false);
  const [showTypeDropdown, setShowTypeDropdown] = useState<boolean>(false);
  
  const searchInput = useRef(null);

  /**
   * Updates the controlled search input's value
   * Enforces "Code" type expectation of 2-3 characters
   * @param event 
   */
  const updateSearchValue = (event: React.ChangeEvent<HTMLInputElement>) => {
    event.preventDefault();

    let newValue = event.target.value;
    if (searchType === "code" && newValue.length > 3) {
      newValue = newValue.slice(0, 3);
    }
    if (submitError) setSubmitError(false);

    setSearchValue(newValue);
  }

  /**
   * Clears the search input's value when switching search types
   * @param event 
   */
  const switchType = (type: SearchType) => {
    if (type === searchType) return;
    if (
      (searchType !== "code" && type === "code") ||
      (type !== "code" && searchType === "code")
    ) {
      setSearchValue("");
    }
    searchInput.current.focus();
    setSearchType(type);
  }

  const toggleTypeDropdown = (visible?: boolean) => {
    if (visible === undefined) {
      setShowTypeDropdown(!showTypeDropdown);
    } else {
      setShowTypeDropdown(visible);
    }
  }

  const handleSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    if (!searchValue.length) {
      setSubmitError(true);
      searchInput.current.focus();
      return;
    }

    fetchCountries(searchType, searchValue);
  }

  // Show red error border when trying to submit if input is empty
  // and lower width when type = "code"
  const formClasses = `search_form search_form__${page}` +
   (submitError ? " error" : "") +
   (searchType === "code" ? " code" : "")

  // Show the clear button only when text is entered
  const clearButtonClasses = "search_form__clear search_form__button" + 
    (searchValue.length ? "" : " hide");
  const searchButtonClasses = "search_form__submit search_form__button" +
    (searchValue.length ? " highlight" : "");

  return (
    <form id={`search_form__${page}`} className={formClasses} onSubmit={handleSubmit}>
      <input 
        id={`search_form__input_${page}`}
        className="search_form__input"
        tabIndex={1} 
        value={searchValue} 
        autoComplete="off" 
        autoCapitalize="off" 
        autoCorrect="off" 
        onChange={updateSearchValue}
        autoFocus={page === "main"}
        ref={searchInput}
      />
      <button 
        id={`search_form__clear`}
        className={clearButtonClasses}
        type="button"
        onClick={() => setSearchValue("")}
      >
        <i className="material-icons">clear</i>
      </button>
      <div className="search_form__inputs">
        <button 
          id={`search_form__type-button_${page}`}
          className="search_form__type-button search_form__button"
          type="button"
          onClick={() => toggleTypeDropdown()}
        >
          <span>{searchType === "fullName" ? "full name" : searchType}</span>
          <i className="material-icons">expand_more</i>
        </button>
        <SearchTypeDropdown 
          page={page}
          visible={showTypeDropdown}
          setVisible={toggleTypeDropdown}
          selectedType={searchType}
          switchType={switchType}
        />
        <button
          id={`search_form__submit_${page}`}
          className={searchButtonClasses}
        >
          <i className="material-icons">search</i>
        </button>
      </div>
      {submitError && (<span className="search_form__error">
        Please enter a value.
      </span>)}
    </form>
  );
}

export default SearchBar;