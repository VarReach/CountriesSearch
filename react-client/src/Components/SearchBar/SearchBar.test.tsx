import React from "react";
import { render, fireEvent, screen } from "@testing-library/react";
// Context
import { SearchContext } from "../../App";
// Components
import SearchBar, { SearchBarProps } from "./SearchBar";
import { SearchType } from "../../Services/CountriesService";

/**
 * Covers tests for the SearchBar and SearchTypeDropdown
 */
describe.only("<SearchBar/>", () => {
  let defaultContext;

  const defaultProps: SearchBarProps = {
    page: "main",
  };

  beforeEach(() => {
    defaultContext = {
      fetchCountries: jest.fn(),
      searchValue: "",
      setSearchValue: jest.fn((value) => defaultContext.searchValue = value),
      searchType: "name" as SearchType,
      setSearchType: jest.fn((type) => defaultContext.searchType = type),
    };
  })

  test("renders without crashing", () => {
    const { container } = render(
      <SearchContext.Provider value={defaultContext}>
        <SearchBar {...defaultProps} />
      </SearchContext.Provider>
    );
    expect(container).toMatchSnapshot();
  });

  describe("<SearchTypeDropdown />", () => {
    test("dropdown should be visible after clicking on button", () => {
      render(
        <SearchContext.Provider value={defaultContext}>
          <SearchBar {...defaultProps} />
        </SearchContext.Provider>
      );

      expect(screen.queryByRole("listbox")).toBeFalsy();
      fireEvent.click(screen.getByRole("button", { name: "name expand_more" }));
      expect(screen.getByRole("listbox")).toBeTruthy();
    });

    test("clicking elsewhere should close the dropdown", () => {
      const { getByRole } = render(
        <SearchContext.Provider value={defaultContext}>
          <SearchBar {...defaultProps} />
        </SearchContext.Provider>
      );

      fireEvent.click(getByRole("button", { name: "name expand_more" }));
      expect(screen.getByRole("listbox")).toBeTruthy();

      fireEvent.click(document);
      expect(screen.queryByRole("listbox")).toBeFalsy();
    });

    test("clicking 'escape' should close the dropdown", () => {
      render(
        <SearchContext.Provider value={defaultContext}>
          <SearchBar {...defaultProps} />
        </SearchContext.Provider>
      );

      fireEvent.click(screen.getByRole("button", { name: "name expand_more" }));
      expect(screen.getByRole("listbox")).toBeTruthy();

      fireEvent.keyDown(screen.getByRole("listbox"), { key: "Escape" });
      expect(screen.queryByRole("listbox")).toBeFalsy();
    });

    test("clicking on a list button should close the dropdown and update search type", () => {
      const { rerender } = render(
        <SearchContext.Provider value={defaultContext}>
          <SearchBar {...defaultProps} />
        </SearchContext.Provider>
      );

      fireEvent.click(screen.getByRole("button", { name: "name expand_more" }));
      expect(screen.getByRole("listbox")).toBeTruthy();

      fireEvent.click(screen.getByText("full name"));

      rerender(
        <SearchContext.Provider value={defaultContext}>
          <SearchBar {...defaultProps} />
        </SearchContext.Provider>
      );
      expect(screen.queryByRole("listbox")).toBeFalsy();
      expect(screen.getByRole("button", { name: "full name expand_more" })).toBeTruthy();
    });

    test("clicking on the selected type should close the dropdown", () => {
      const { rerender } = render(
        <SearchContext.Provider value={defaultContext}>
          <SearchBar {...defaultProps} />
        </SearchContext.Provider>
      );

      fireEvent.click(screen.getByRole("button", { name: "name expand_more" }));
      expect(screen.getByRole("listbox")).toBeTruthy();

      const elements = screen.getAllByText("name");
      fireEvent.click(elements[1]);
      expect(screen.queryByRole("listbox")).toBeFalsy();
      expect(screen.getByRole("button", { name: "name expand_more" })).toBeTruthy();
    });
  });

  test("When search type is code, the input is limited to max 3 characters", () => {
    const { rerender } = render(
      <SearchContext.Provider value={defaultContext}>
        <SearchBar {...defaultProps} />
      </SearchContext.Provider>
    );

    fireEvent.click(screen.getByRole("button", { name: "name expand_more" }));
    expect(screen.getByRole("listbox")).toBeTruthy();

    fireEvent.click(screen.getByText("code"));

    rerender(
      <SearchContext.Provider value={defaultContext}>
        <SearchBar {...defaultProps} />
      </SearchContext.Provider>
    );

    expect(screen.queryByRole("listbox")).toBeFalsy();
    expect(screen.getByRole("button", { name: "code expand_more" })).toBeTruthy();

    fireEvent.change(screen.getByRole("textbox"), { target: { value: "testing" } });

    rerender(
      <SearchContext.Provider value={defaultContext}>
        <SearchBar {...defaultProps} />
      </SearchContext.Provider>
    );

    expect(screen.queryByDisplayValue("testing")).toBeFalsy();
    expect(screen.getByDisplayValue("tes")).toBeTruthy();
  });

  describe("Switching to/from type Code should clear input", () => {
    test("switching from name to code", () => {
      render(
        <SearchContext.Provider value={defaultContext}>
          <SearchBar {...defaultProps} />
        </SearchContext.Provider>
      );

      fireEvent.change(screen.getByRole("textbox"), { target: { value: "testing" } });
      fireEvent.click(screen.getByRole("button", { name: "name expand_more" }));
      fireEvent.click(screen.getByText("code"));
  
      expect(screen.queryByDisplayValue("testing")).toBeFalsy();
    });

    test("switching from full name to code", () => {
      render(
        <SearchContext.Provider value={defaultContext}>
          <SearchBar {...defaultProps} />
        </SearchContext.Provider>
      );
      // Switch to fullname
      fireEvent.click(screen.getByRole("button", { name: "name expand_more" }));
      fireEvent.click(screen.getByText("full name"));

      fireEvent.change(screen.getByRole("textbox"), { target: { value: "testing" } });

      // Switch to code
      fireEvent.click(screen.getByRole("button", { name: "full name expand_more" }));
      fireEvent.click(screen.getByText("code"));
  
      expect(screen.queryByDisplayValue("testing")).toBeFalsy();
    });

    test("switching from code to name", () => {
      render(
        <SearchContext.Provider value={defaultContext}>
          <SearchBar {...defaultProps} />
        </SearchContext.Provider>
      );
      // Switch to fullname
      fireEvent.click(screen.getByRole("button", { name: "name expand_more" }));
      fireEvent.click(screen.getByText("code"));

      fireEvent.change(screen.getByRole("textbox"), { target: { value: "tes" } });

      // Switch to code
      fireEvent.click(screen.getByRole("button", { name: "code expand_more" }));
      fireEvent.click(screen.getByText("name"));
  
      expect(screen.queryByDisplayValue("tes")).toBeFalsy();
    });

    test("switching from code to full name", () => {
      render(
        <SearchContext.Provider value={defaultContext}>
          <SearchBar {...defaultProps} />
        </SearchContext.Provider>
      );
      // Switch to fullname
      fireEvent.click(screen.getByRole("button", { name: "name expand_more" }));
      fireEvent.click(screen.getByText("code"));

      fireEvent.change(screen.getByRole("textbox"), { target: { value: "tes" } });

      // Switch to code
      fireEvent.click(screen.getByRole("button", { name: "code expand_more" }));
      fireEvent.click(screen.getByText("full name"));
  
      expect(screen.queryByDisplayValue("tes")).toBeFalsy();
    });
  });

  test("should not attempt to fetch countries if input is empty", () => {
    render(
      <SearchContext.Provider value={defaultContext}>
        <SearchBar {...defaultProps} />
      </SearchContext.Provider>
    );

    fireEvent.click(screen.getByRole("button", { name: "name expand_more" }));
    expect(screen.getByRole("listbox")).toBeTruthy();

    fireEvent.click(screen.getByRole("button", { name: "search" }));
    expect(defaultContext.fetchCountries).not.toHaveBeenCalled();
  });

  test("should attempt to fetch countries if input is empty", () => {
    const { rerender } = render(
      <SearchContext.Provider value={defaultContext}>
        <SearchBar {...defaultProps} />
      </SearchContext.Provider>
    );

    fireEvent.click(screen.getByRole("button", { name: "name expand_more" }));
    expect(screen.getByRole("listbox")).toBeTruthy();

    fireEvent.change(screen.getByRole("textbox"), { target: { value: "testing" } });
    rerender(
      <SearchContext.Provider value={defaultContext}>
        <SearchBar {...defaultProps} />
      </SearchContext.Provider>
    );

    fireEvent.click(screen.getByRole("button", { name: "search" }));

    expect(defaultContext.fetchCountries).toHaveBeenCalled();
  });
});