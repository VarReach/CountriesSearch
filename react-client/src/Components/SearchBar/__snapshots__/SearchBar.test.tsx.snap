// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchBar/> renders without crashing 1`] = `
<div>
  <form
    class="search_form search_form__main"
    id="search_form__main"
  >
    <input
      autocapitalize="off"
      autocomplete="off"
      autocorrect="off"
      class="search_form__input"
      id="search_form__input_main"
      tabindex="1"
      value=""
    />
    <button
      class="search_form__clear search_form__button hide"
      id="search_form__clear"
      type="button"
    >
      <i
        class="material-icons"
      >
        clear
      </i>
    </button>
    <div
      class="search_form__inputs"
    >
      <button
        class="search_form__type-button search_form__button"
        id="search_form__type-button_main"
        type="button"
      >
        <span>
          name
        </span>
        <i
          class="material-icons"
        >
          expand_more
        </i>
      </button>
      <button
        class="search_form__submit search_form__button"
        id="search_form__submit_main"
      >
        <i
          class="material-icons"
        >
          search
        </i>
      </button>
    </div>
  </form>
</div>
`;
